### DISCUSSION TITLE
Run Custom S6 Script as root

### DISCUSSION THREAD

**Discussion by papoms** [serversideup/docker-php#425] [Q&A] [answered]
I am successfully using custom s6 services to run stuff in the container.

Now with the switch from v2 -> v3 i am facing an issue where i used a oneshot script to install something as root (in v2).

How would i "sudo" here to install as root?

_/etc/s6-overlay/s6-rc.d/install-software-x_
```
#!/command/with-contenv bash
/install_xyz.sh
```

**Comment by jaydrogers** [comment, marked as answer]
`sudo` is not available by default in our images. So I see two options here:

1. Run our images as root (Add `USER root`) at the end of your image (like we do for the `ci` step here https://serversideup.net/open-source/docker-php/docs/guide/migrating-from-v2-to-v3#cicd)
2. Install and configure sudo and grant www-data to do root things (which might be worse than just running the image as `root` because at least `root` will start services as `www-data` and anonymous users will not have access to do `root` things)

Hope this helps!

**Reply by LorenzoRogai** [reply]
@jaydrogers on the latest images, this error appears when `USER root` is left at the end of the Dockerfile

Is this normal? The image still proceeds to boot and run normally afterward

```
/package/admin/s6-overlay/libexec/preinit: info: /run belongs to uid 33 instead of 0 - fixing it

[29-Jan-2026 15:20:49] ALERT: [pool www] user has not been defined

[29-Jan-2026 15:20:49] ERROR: failed to post process the configuration

[29-Jan-2026 15:20:49] ERROR: FPM initialization failed
```

**Reply by jaydrogers** [reply]
This message is okay:

> /package/admin/s6-overlay/libexec/preinit: info: /run belongs to uid 33 instead of 0 - fixing it

It's just a warning and it fixes itself.

These I cannot replicate:

> [29-Jan-2026 15:20:49] ALERT: [pool www] user has not been defined
> [29-Jan-2026 15:20:49] ERROR: failed to post process the configuration
> [29-Jan-2026 15:20:49] ERROR: FPM initialization failed

Command to replicate:

```
docker run --rm --pull always --user root serversideup/php:8.5-fpm-nginx
```

Results:

```
8.5-fpm-nginx: Pulling from serversideup/php
Digest: sha256:14724e05e959da090b061ed445d0f73568bb5e69f1b1b3a694f931a619a964b3
Status: Image is up to date for serversideup/php:8.5-fpm-nginx

--------------------------------------------------------------------
 ____                             ____  _     _        _   _
/ ___|  ___ _ ____   _____ _ __  / ___|(_) __| | ___  | | | |_ __
\___ \ / _ \  __\ \ / / _ \  __| \___ \| |/ _` |/ _ \ | | | |  _ \
 ___) |  __/ |   \ V /  __/ |     ___) | | (_| |  __/ | |_| | |_) |
|____/ \___|_|    \_/ \___|_|    |____/|_|\__,_|\___|  \___/| .__/
                                                            |_|

Brought to you by serversideup.net
--------------------------------------------------------------------

📚 Documentation: https://serversideup.net/php/docs
💬 Get Help: https://serversideup.net/php/community
🙌 Sponsor: https://serversideup.net/sponsor

-------------------------------------
ℹ️ Container Information
-------------------------------------
📦 Versions
• Image:         v4.3.0-21389458876
• PHP:           8.5.2
• OS:            Debian GNU/Linux 13 (trixie)

👤 Container User
• User:          root
• UID:           0
• GID:           0

⚡ Performance
• OPcache:       ❌ Disabled
• Memory Limit:  256M
• Upload Limit:  100M

🔄 Runtime
• Docker CMD:     /init

👉 [NOTICE]: Improve PHP performance by setting PHP_OPCACHE_ENABLE=1 (recommended for production).
ℹ️ NOTICE (generate-ssl): SSL mode is off, so we won't generate a self-signed SSL key pair.
(init-webserver-config): Processing /etc/nginx/nginx.conf.template → /etc/nginx/nginx.conf...
(init-webserver-config): Processing /etc/nginx/site-opts.d/http.conf.template → /etc/nginx/site-opts.d/http.conf...
(init-webserver-config): Processing /etc/nginx/site-opts.d/https.conf.template → /etc/nginx/site-opts.d/https.conf...
(init-webserver-config): Processing /etc/nginx/sites-available/ssl-full.template → /etc/nginx/sites-available/ssl-full...
ℹ️ NOTICE (init-webserver-config): Enabling NGINX site with SSL "off"...
/package/admin/s6-overlay/libexec/preinit: info: /run belongs to uid 33 instead of 0 - fixing it
[29-Jan-2026 16:52:26] NOTICE: fpm is running, pid 112
[29-Jan-2026 16:52:26] NOTICE: ready to handle connections
✅ NGINX + PHP-FPM is running correctly.
```

**Reply by LorenzoRogai** [reply]
@jaydrogers it happens when using docker-php-serversideup-s6-init

```
FROM serversideup/php:8.4-fpm-nginx

ENV S6_CMD_WAIT_FOR_SERVICES=1

ENV AUTORUN_ENABLED "true"
ENV AUTORUN_LARAVEL_MIGRATION "false"
ENV PHP_OPCACHE_ENABLE "1"

ENV PHP_MEMORY_LIMIT "4096M"

# Register custom startup scripts
COPY --chmod=755 ./entrypoint.d/ /etc/entrypoint.d/

# Change to root, so we can do root things
USER root

# As root, run the docker-php-serversideup-s6-init script
RUN docker-php-serversideup-s6-init
```

**Reply by LorenzoRogai** [reply]
@jaydrogers did you have a chance to look into this? We need to run one service as root (wazuh-agent), but the error is still occurring if we remove ```USER www-data``` at the end of the Dockerfile

If you build the above Dockerfile and run it those errors will appear

EDIT: I created a PR for this, i already tested it locally and it works https://github.com/serversideup/docker-php/pull/689/changes
